hri_safety_sense/src/VscProcess.cpp

```diff
@@ -15,207 +15,240 @@
-	vscInterface = vsc_initialize(serialPort.c_str(),serialSpeed);
-	if (vscInterface == NULL) {
-		ROS_FATAL("Cannot open serial port! (%s, %i)",serialPort.c_str(),serialSpeed);
-	} else {
-		ROS_INFO("Connected to VSC on %s : %i",serialPort.c_str(),serialSpeed);
-	}
-
-	// Attempt to Set priority
-	bool  set_priority = false;
-	if(nh.getParam("set_priority", set_priority)) {
-		ROS_INFO("Set priority updated to:  %i",set_priority);
-	}
-
-	if(set_priority) {
-		if(setpriority(PRIO_PROCESS, 0, -19) == -1) {
-			ROS_ERROR("UNABLE TO SET PRIORITY OF PROCESS! (%i, %s)",errno,strerror(errno));
-		}
-	}
-
-	// Create Message Handlers
-	joystickHandler = new JoystickHandler(frameId);
-
-	// EStop callback
-	estopServ = rosNode.advertiseService("safety/service/send_emergency_stop", &VscProcess::EmergencyStop, this);
-
-	// KeyValue callbacks
-	keyValueServ = rosNode.advertiseService("safety/service/key_value", &VscProcess::KeyValue, this);
-	keyStringServ = rosNode.advertiseService("safety/service/key_string", &VscProcess::KeyString, this);
-
-	// Publish Emergency Stop Status
-	estopPub = rosNode.advertise<std_msgs::UInt32>("safety/emergency_stop", 10);
-
-	// Main Loop Timer Callback
-	mainLoopTimer = rosNode.createTimer(ros::Duration(1.0/VSC_INTERFACE_RATE), &VscProcess::processOneLoop, this);
-
-	// Init last time to now
-	lastDataRx = ros::Time::now();
-
-	// Clear all error counters
-	memset(&errorCounts, 0, sizeof(errorCounts));
+  std::string serialPort = this->declare_parameter<std::string>("serial",
+    "/dev/ttyACM0");
+  RCLCPP_INFO(this->get_logger(), "Serial Port set to:  %s", serialPort.c_str());
+
+  int serialSpeed = this->declare_parameter<int>("serial_speed", 115200);
+  RCLCPP_INFO(this->get_logger(), "Serial Port Speed set to:  %i", serialSpeed);
+
+  std::string frameId = this->declare_parameter<std::string>("frame_id",
+    "/srcs");
+  RCLCPP_INFO(this->get_logger(), "Frame ID set to:  %s", frameId.c_str());
+
+  /* Open VSC Interface */
+  vscInterface_ = vsc_initialize(serialPort.c_str(), serialSpeed);
+  if (vscInterface_ == NULL) {
+    RCLCPP_FATAL(this->get_logger(), "Cannot open serial port! (%s, %i)", serialPort.c_str(), serialSpeed);
+    throw std::system_error(std::make_error_code(std::errc::io_error), "Cannot open serial port");
+  } else {
+    RCLCPP_INFO(this->get_logger(), "Connected to VSC on %s : %i", serialPort.c_str(), serialSpeed);
+  }
+
+  // Attempt to Set priority
+  bool setPriority = this->declare_parameter<bool>("set_priority", false);
+  RCLCPP_INFO(this->get_logger(), "Set priority:  %i", setPriority);
+
+  if (setPriority) {
+    if (setpriority(PRIO_PROCESS, 0, -19) == -1) {
+      RCLCPP_ERROR(this->get_logger(), "Unable to set priority of process! (%i, %s)", errno, strerror(errno));
+      throw std::system_error(errno, std::system_category(), strerror(errno));
+    }
+  }
+
+  // Create Message Handlers
+  joystickHandler_ = new JoystickHandler(this->get_node_topics_interface(),
```

Comment:
Instead of using `new` here, let's switch to modern C++ smart pointers.  Since I don't think you are sharing this pointer with any other class, switching this to a `std::unique_ptr` seems to be the right way to go.  Don't forget to remove the delete in the destructor.

Reply:
d580097